### ISSUE TITLE
Logger connector callback metrics only contains the last step

### ISSUE THREAD

**Issue by ZhaofengWu** [PyTorchLightning/pytorch-lightning#13147]
## 🐛 Bug

When I do
```
def training_step(...):
   self.log("train_los", ...)
```
and see that `log()` by default uses mean reduction, I expect the `"train_loss"` given back to me in the callback metrics in the logger connector to be the average training loss across all examples. But right now, this is actually the training loss of the last batch, because lightning considers this to be an `on_step` metric. I argue that this is very unintuitive (and also undocumented). This is especially problematic when this metric is used to, e.g., perform epoch selection, since the performance on a single batch can have large variance. Worse, each epoch can have a different batch order, and hence a different last batch, so the loss value isn't really comparable if this is the case.

If this is a design decision, could you at least let me know how I could achieve my intended purpose?

cc @carmocca @edward-io @ananthsub @rohitgr7 @kamil-kaczmarek @Raalsky @Blaizzy

**Comment by carmocca** [comment]
I believe you want to set:

```python
def training_step(...):
    loss = ...
    self.log("train_loss", loss, on_step=False, on_epoch=True)
```

This is described in the docs here: https://pytorch-lightning.readthedocs.io/en/stable/extensions/logging.html#logging-from-a-lightningmodule

**Comment by ZhaofengWu** [comment]
Thanks, this is what I thought as well. But I argue that this is not a very intuitive behavior due to what I mentioned above. A user might provide `reduce_fx=torch.mean`, but this argument would be silently ignored. I feel like it's better to assert that the reduce function must be a function that takes the last element in a sequence/tensor, or None, or something like that.

Anyway, thanks for helping confirm this, and feel free to close this issue if you prefer to keep the current behavior.

**Comment by carmocca** [comment]
I think I'm missing your point.

By default, `log` does use "mean" as the `reduce_fx`. Why do you say it is silently ignored?
Note that it only supports mean, min, max, sum. Support for any other reductions is not planned and, if needed, we suggest logging a `torchmetrics.Metric` instead.

This was recently clarified with https://github.com/PyTorchLightning/pytorch-lightning/pull/13101

**Comment by ZhaofengWu** [comment]
Suppose a user calls `self.log("train_loss", loss, reduce_fx=torch.mean)` (which happens to be the default, but writing it out be clearer), I think the expected behavior is that when this metric is logged, e.g. in `callback_metrics`, this aggregate reduced value would be the mean. But instead, this `reduce_fx` is not used at all (i.e., silently ignored), and under the hood the last value is considered the aggregate value. I think this is a bit unintuitive.

**Comment by ZhaofengWu** [comment]
Also, when I do `self.log("train_loss", loss, on_step=False, on_epoch=True)` as you suggested, `train_loss` is not returned in `trainer.callback_metrics` in `on_validation_end` of my logging callback when the first epoch ends, but it is only returned starting from when the second epoch ends. Any idea why this is happening?

**Comment by carmocca** [comment]
> his aggregate reduced value would be the mean. But instead, this reduce_fx is not used at all (i.e., silently ignored), and under the hood the last value is considered the aggregate value. I think this is a bit unintuitive.

Logging in `training_step` defaults to `on_step=True, on_epoch=False` and that setting produces the behaviour you are seeing, where the values are not aggregated but logged at each step instead. For that, you want `on_step=False, on_epoch=True`

> Any idea why this is happening?

This happens because in the current design, if you access `callback_metrics` during validation, it will only contain the validation results.
You can work around this by accessing  `callback_metrics` in `on_train_epoch_end` instead which will be executed right after `on_validation_end`

**Comment by ZhaofengWu** [comment]
> Logging in training_step defaults to on_step=True, on_epoch=False and that setting produces the behaviour you are seeing, where the values are not aggregated but logged at each step instead. For that, you want on_step=False, on_epoch=True

Right, I now understand what the correct way to do this is, but I'm just arguing that this is perhaps a bit unintuitive. Anyway, this is not a very big point.

> This happens because in the current design, if you access callback_metrics during validation, it will only contain the validation results. You can work around this by accessing callback_metrics in on_train_epoch_end instead which will be executed right after on_validation_end

I see, thanks for explaining! So if I want to print the average training loss and the validation metrics after each epoch (with the traditional setup that one validation is done at the end of each training epoch), is the following the right way to do it?

1. In `LightningModule.training_step`, do `self.log("train_loss", loss, on_step=False, on_epoch=True)`;
2. In `LightningModule.validation_epoch_end`, do `self.log("validation_metric", validation_metric)`; 
3. In  logging `Callback.on_validation_end`, print `trainer.callback_metrics["validation_metric"]`;
4. In  logging `Callback.on_train_epoch_end`, print `trainer.callback_metrics["train_loss"]`.

And what if I want to run validation multiple times during one epoch? Is it right to say that it's impossible to report the current training loss together with the validation metrics?

**Comment by carmocca** [comment]
You can skip step (3) and print both `validation_metric` and `train_loss` in `on_train_epoch_end`. This will also work when validation runs multiple times per epoch

**Comment by ZhaofengWu** [comment]
Got it, thanks a lot!
